squidpy/read/_read.py

```diff
@@ -0,0 +1,197 @@
+    }
+    image_dic = _read_images(image_files)
+
+    adata.uns[Key.uns.spatial][library_id][Key.uns.image_key] = image_dic
+
+    coords_scalefactors_files = {
+        "tissue_positions_file": path / f"{Key.uns.spatial}/tissue_positions_list.csv",
+        "scalefactors_json_file": path / f"{Key.uns.spatial}/scalefactors_json.json",
+    }
+
+    adata.uns[Key.uns.spatial][library_id]["scalefactors"] = json.loads(
+        coords_scalefactors_files["scalefactors_json_file"].read_bytes()
+    )
+
+    columns = [
+        "barcode",
+        "in_tissue",
+        "array_row",
+        "array_col",
+        "pxl_col_in_fullres",
+        "pxl_row_in_fullres",
+    ]
+
+    coords = _read_coords(
+        coords_scalefactors_files["tissue_positions_file"],
+        adata.shape[0],
+        columns,
+        kwargs={"header": None},
+    )
+
+    coords.index = coords["barcode"].values  # type: ignore
+
+    adata.obs = adata.obs.join(coords, how="left")
+
+    adata.obsm[Key.uns.spatial] = adata.obs[["pxl_row_in_fullres", "pxl_col_in_fullres"]].to_numpy()
+    adata.obs.drop(
+        columns=["barcode", "pxl_row_in_fullres", "pxl_col_in_fullres"],
+        inplace=True,
+    )
+
+    # put image path in uns
+    if source_image_path is not None:
+        # get an absolute path
+        source_image_path = str(Path(source_image_path).resolve())
+        adata.uns[Key.uns.spatial][library_id]["metadata"]["source_image_path"] = str(source_image_path)
+    return adata
+
+
+# def _read_nanostring(
+#     path: str | Path,
+#     genome: str | None,
+#     *,
+#     count_file: str = "",
+#     library_id: str | None = None,
+#     load_images: bool | None = True,
+#     source_image_path: str | Path | None = None,
+# ) -> AnnData:
+
+
+def read_vizgen(
+    path: str | Path,
+    *,
+    count_file: str,
+    obs_file: str,
+    library_id: str | None = None,
+) -> AnnData:
+    r"""
+    Read Vizgen formatted dataset.
+
+    In addition to reading the regular Vizgen output,
+    this function loads metadata file and loads the spatial co-ordinates.
+    """
+    if isinstance(path, str):
+        path = Path(path)
+
+    text_kwargs = {
+        "first_column_names": True,
+        "delimiter": ",",
+    }
+
+    adata = _read_count(path=path, count_file=count_file, text_kwargs=text_kwargs)
+
+    if library_id is not None:
+        adata.uns[Key.uns.spatial] = {library_id: {}}
+    else:
+        adata.uns[Key.uns.spatial] = {}
+
+    coords_file = {
+        "metadata_file": path / obs_file,
+    }
+
+    columns = [
+        "fov",
+        "volume",
+        "center_x",
+        "center_y",
+        "min_x",
+        "max_x",
+        "min_y",
+        "max_y",
+    ]
+
+    coords = _read_coords(
+        coords_file["metadata_file"],
+        n_obs=adata.shape[0],
+        cols=columns,
+        header=0,
+        index_col=0,
+    )
+
+    adata.obs = adata.obs.join(coords, how="left")
```

Comment:
Could `adata` ever have a different ordering than `coords`? Because this call will reorder the `obs` dataframe, while your assignment to `.obsm` below will not.